BasicRecordingApiKotlin/app/src/main/java/com/google/android/gms/fit/samples/basicrecordingapikotlin/MainActivity.kt

```diff
@@ -0,0 +1,326 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.android.gms.fit.samples.basicrecordingapikotlin
+
+import android.Manifest
+import android.content.Intent
+import android.content.pm.PackageManager
+import android.graphics.Color
+import android.net.Uri
+import android.os.Build
+import android.os.Bundle
+import android.provider.Settings
+import android.view.Menu
+import android.view.MenuItem
+import android.view.View
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.app.ActivityCompat
+import androidx.core.widget.TextViewCompat
+import com.google.android.gms.auth.api.signin.GoogleSignIn
+import com.google.android.gms.fit.samples.common.logger.Log
+import com.google.android.gms.fit.samples.common.logger.LogView
+import com.google.android.gms.fit.samples.common.logger.LogWrapper
+import com.google.android.gms.fit.samples.common.logger.MessageOnlyLogFilter
+import com.google.android.gms.fitness.Fitness
+import com.google.android.gms.fitness.FitnessOptions
+import com.google.android.gms.fitness.data.DataType
+import com.google.android.material.snackbar.Snackbar
+
+const val TAG = "BasicRecordingApi"
+
+/**
+ * This sample demonstrates how to use the Recording API of the Google Fit platform to subscribe
+ * to data sources, query against existing subscriptions, and remove subscriptions. It also
+ * demonstrates how to authenticate a user with Google Play Services.
+ *
+ * Because the subscription is for data type TYPE_CALORIES_EXPENDED, on Android 10 the Android
+ * ACTIVITY_RECOGNITION permission is required. This sample shows how to check for and request the
+ * permission when necessary.
+ */
+class MainActivity : AppCompatActivity() {
+    private val fitnessOptions: FitnessOptions by lazy {
+        FitnessOptions.builder()
+                .addDataType(DataType.TYPE_CALORIES_EXPENDED, FitnessOptions.ACCESS_WRITE)
+                .build()
+    }
+
+    // Two maps are used in conjunction with the OAuth flow. This is in order to identify what
+    // function to execute once the flow has successfully completed.
+    private val signInActionMap = mapOf<() -> Any, Int>(
+            ::subscribe to 1,
+            ::dumpSubscriptionsList to 2,
+            ::cancelSubscription to 3
+    )
+    private val signInResponseMap = signInActionMap.entries.associate { (k, v) -> v to k }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        // This method sets up our custom logger, which will print all log messages to the device
+        // screen, as well as to adb logcat.
+        initializeLogging()
+
+        checkPermissionsAndRun(::subscribe)
+    }
+
+    private fun checkPermissionsAndRun(signInAction: () -> Any) {
+        if (!requiresActivityRecognitionPermission() || hasRuntimePermission()) {
```

Comment:
Rename hasRuntimePermission() to permissionApproved(), it's much easier to read. I would also put that as the first argument in the OR.

Reply:
Actually, a second comment which you could apply to the other samples. Why does this Q or not need to be separate from permissionsApproved(), can't it be part of it? It makes this statement much harder to read. I would copy/paste the code I wrote for the location permission codelab and use that (link below). It hides the Q stuff... of course you want to rename it since this isn't foreground... it uses another method above it as a helper. Let me know if you have any questions.

https://github.com/googlecodelabs/while-in-use-location/blob/master/complete/src/main/java/com/example/android/whileinuselocation/MainActivity.kt#L219